File: src/runtime/variables.ts
export class VariableStore {
  private readonly values = new Map<string, unknown>();

  set(name: string, value: unknown): void {
    this.values.set(name, value);
  }

  get(name: string): unknown {
    return this.values.get(name);
  }

  has(name: string): boolean {
    return this.values.has(name);
  }

  snapshot(): Record<string, unknown> {
    return Object.fromEntries(this.values.entries());
  }
}

File: src/runtime/variables.rs
use std::collections::BTreeMap;

use serde_json::Value;

#[derive(Debug, Default, Clone)]
pub struct VariableStore {
    values: BTreeMap<String, Value>,
}

impl VariableStore {
    pub fn set(&mut self, name: impl Into<String>, value: Value) {
        self.values.insert(name.into(), value);
    }

    pub fn get(&self, name: &str) -> Option<&Value> {
        self.values.get(name)
    }

    pub fn has(&self, name: &str) -> bool {
        self.values.contains_key(name)
    }

    pub fn snapshot(&self) -> BTreeMap<String, Value> {
        self.values.clone()
    }
}
